### PR TITLE
arm64: dts: qcom: samsung-gtelwifiue: Add panel and touchscreen

### PR DESCRIPTION
**Pull request by person4268**
Adds the generated panel driver for the device, and adds it to the
dtb, along with support for the ZT7554 touchscreen. The touchscreen
works fine, along with the panel.

Signed-off-by: Michael Abood <person4265@gmail.com>

### CODE REVIEW

**Comment by person4268** [comment]
Converting to draft  for now since build is failing because the generated panel driver is producing a warning

File: arch/arm64/boot/dts/qcom/apq8016-samsung-gtelwifiue.dts [hunk 1]
```diff
@@ -18,14 +18,6 @@
 		stdout-path = "serial0";
 	};
 
-	reserved-memory {
```
Comment (TravMurav) [inline arch/arm64/boot/dts/qcom/apq8016-samsung-gtelwifiue.dts:21]: Please don't move blocks back and forth - this makes patches harder to apply out of order.

File: arch/arm64/boot/dts/qcom/apq8016-samsung-gtelwifiue.dts [hunk 2]
```diff
@@ -72,10 +85,52 @@
...
+		compatible = "zinitix,zt7554", "zinitix,bt541";
+		reg = <0x20>;
+		interrupt-parent = <&msmgpio>;
+		interrupts = <13 IRQ_TYPE_EDGE_FALLING>;
+
+		touchscreen-size-x = <799>;
+		touchscreen-size-y = <1279>;
+		linux,keycodes = <KEY_APPSELECT KEY_BACK>;
+
+		vdd-supply = <&reg_vdd_tsp>;
```
Comment (TravMurav) [inline arch/arm64/boot/dts/qcom/apq8016-samsung-gtelwifiue.dts:102]: Consider using the new supply names as documented in the future schema:

https://github.com/TravMurav/linux/blob/424e1247eb374b4495c9e46304fc85f3345e8149/Documentation/devicetree/bindings/input/touchscreen/zinitix,bt400.yaml#L55-L62

File: arch/arm64/boot/dts/qcom/apq8016-samsung-gtelwifiue.dts [hunk 3]
```diff
@@ -72,10 +85,52 @@
...
+		pinctrl-0 = <&tsp_int_default>;
+	};
+};
+
 &blsp1_uart2 {
 	status = "okay";
 };
 
+&dsi0 {
+        panel@0 {
```
Comment (TravMurav) [inline arch/arm64/boot/dts/qcom/apq8016-samsung-gtelwifiue.dts:115]: Consider configuring your editor to use tabs (see checkpatch errors in CI)

File: arch/arm64/boot/dts/qcom/apq8016-samsung-gtelwifiue.dts [hunk 4]
```diff
@@ -72,10 +85,52 @@
 	};
 };
 
+&blsp_i2c5 {
+	status = "okay";
+
+	// Real IC is ZT7554 but BT541 appears to work just fine
+	touchscreen@20 {
```
Comment (TravMurav) [inline arch/arm64/boot/dts/qcom/apq8016-samsung-gtelwifiue.dts:92]: Please separate this patch into two: One introduces touch support, another adds panel. (One of the reasons for that is that the panel isn't possible to upstream right now but the touch can be)

You can use `git add --patch` to split the changes into hunks

See e.g. https://github.com/msm8916-mainline/linux/commits/master/arch/arm64/boot/dts/qcom/msm8916-samsung-gt58.dts

File: drivers/gpu/drm/panel/msm8916-generated/panel-samsung-gtelwifiue-nt51017.c [hunk 5]
```diff
@@ -51,8 +51,6 @@ static int nt51017_on(struct nt51017 *ctx)
 
 static int nt51017_off(struct nt51017 *ctx)
 {
-	struct mipi_dsi_device *dsi = ctx->dsi;
```
Comment (TravMurav) [inline drivers/gpu/drm/panel/msm8916-generated/panel-samsung-gtelwifiue-nt51017.c]: Please make this into another patch.

See e.g. https://github.com/msm8916-mainline/linux/commits/master/drivers/gpu/drm/panel

File: arch/arm64/boot/dts/qcom/apq8016-samsung-gtelwifiue.dts [hunk 6]
```diff
@@ -72,10 +85,52 @@
...
 &blsp1_uart2 {
 	status = "okay";
 };
 
+&dsi0 {
+        panel@0 {
+                compatible = "samsung,gtelwifiue-nt51017";
+                reg = <0>;
+
+                //backlight = <&backlight>; // No backlight for now
```
Comment (TravMurav) [inline arch/arm64/boot/dts/qcom/apq8016-samsung-gtelwifiue.dts:119]:
```suggestion
```

File: arch/arm64/boot/dts/qcom/apq8016-samsung-gtelwifiue.dts (hunk 2 above)
Comment (Minecrell) [reply arch/arm64/boot/dts/qcom/apq8016-samsung-gtelwifiue.dts:102]: I didn't apply the patch that allows using them yet :D

File: drivers/gpu/drm/panel/msm8916-generated/panel-samsung-gtelwifiue-nt51017.c (hunk 5 above)
Comment (Minecrell) [reply drivers/gpu/drm/panel/msm8916-generated/panel-samsung-gtelwifiue-nt51017.c]: Yes, patches should be separated by file (or at least subsystem).

Comment (person4268) [reply drivers/gpu/drm/panel/msm8916-generated/panel-samsung-gtelwifiue-nt51017.c]: Is it fine if they're in the same PR the way they are right now?

File: arch/arm64/boot/dts/qcom/apq8016-samsung-gtelwifiue.dts (hunk 3 above)
Comment (person4268) [reply arch/arm64/boot/dts/qcom/apq8016-samsung-gtelwifiue.dts:115]: Fun, my editor is set to use tabs. I probably carelessly copied it from the terminal and my editor doesn't format pasted text.

File: drivers/gpu/drm/panel/msm8916-generated/panel-samsung-gtelwifiue-nt51017.c (hunk 5 above)
Comment (person4268) [reply drivers/gpu/drm/panel/msm8916-generated/panel-samsung-gtelwifiue-nt51017.c]: Ok, I went and split everything up.

**Comment by person4268** [comment]
Do I need to sign-off the panel PR, or what should I do regarding it? Or should I straight up do something like change the author to me?

File: drivers/gpu/drm/panel/msm8916-generated/panel-samsung-nt51017-b4p096wx5vp09.c [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+
+	if (ctx->prepared)
+		return 0;
+
+	ret = regulator_enable(ctx->supply);
+	if (ret < 0) {
+		dev_err(dev, "Failed to enable regulator: %d\n", ret);
+		return ret;
+	}
+
```
Comment (Minecrell) [inline drivers/gpu/drm/panel/msm8916-generated/panel-samsung-nt51017-b4p096wx5vp09.c:73]: Typically there is some delay here automatically because there is some reset sequence that can be run. Doesn't look like that exists for this panel. But then I suggest we add some delay here, specifically the `samsung,mdss-power-on-reset-delay-us = <30000>;` using

```c
msleep(30);
```

File: arch/arm64/boot/dts/qcom/apq8016-samsung-gtelwifiue.dts [hunk 8]
```diff
@@ -72,10 +102,52 @@
...
 &blsp1_uart2 {
 	status = "okay";
 };
 
+&dsi0 {
+	panel@0 {
+		compatible = "samsung,nt51017-b4p096wx5vp09";
+		reg = <0>;
+
+		led-supply = <&reg_lcd>;
```
Comment (Minecrell) [inline arch/arm64/boot/dts/qcom/apq8016-samsung-gtelwifiue.dts:136]: typo

File: arch/arm64/boot/dts/qcom/apq8016-samsung-gtelwifiue.dts [hunk 9]
```diff
@@ -61,6 +61,20 @@
...
+		regulator-min-microvolt = <3300000>;
+		regulator-max-microvolt = <3300000>;
+
+		gpio = <&msmgpio 73 GPIO_ACTIVE_HIGH>;
+		enable-active-high;
+
+		pinctrl-names = "default";
+		pinctrl-0 = <&reg_tsp_en_default>;
+
 		startup-delay-us = <150>;
```
Comment (TravMurav) [inline arch/arm64/boot/dts/qcom/apq8016-samsung-gtelwifiue.dts:78]: `startup-delay-us` jumped from `regulator-lcd` into this one? (If you intended to drop the line because of the always-on property, it shouldn't appear in the previous commit to begin with, we usually consider patches as separate and independent so reverting/reordering/moving patches across trees should be as clean as possible)

File: arch/arm64/boot/dts/qcom/apq8016-samsung-gtelwifiue.dts [hunk 10]
```diff
@@ -89,6 +103,28 @@
...
+
+	// Real IC is ZT7554 but BT541 appears to work just fine
+	touchscreen@20 {
+		compatible = "zinitix,zt7554", "zinitix,bt541";
+		reg = <0x20>;
+		interrupt-parent = <&msmgpio>;
+		interrupts = <13 IRQ_TYPE_EDGE_FALLING>;
+
+		touchscreen-size-x = <799>;
+		touchscreen-size-y = <1279>;
```
Comment (TravMurav) [inline arch/arm64/boot/dts/qcom/apq8016-samsung-gtelwifiue.dts:117]: Are you really sure that 1280x800 (matching the display) doesn't work there?

**Review by TravMurav — COMMENTED** [review]
I'd also appreciate if the commit adding the touch was the first to be applied (as it's more "ready" compared to the other ones) and reworded in a way as if the correct compatible is used already (I think I will add it to the driver as I already have some pending series on the mailing list, so then we'll just drop the extra compatible)

File: arch/arm64/boot/dts/qcom/apq8016-samsung-gtelwifiue.dts (hunk 9 above)
Comment (Minecrell) [reply arch/arm64/boot/dts/qcom/apq8016-samsung-gtelwifiue.dts:78]: Looks like a well hidden rebase fail to me :D  
The `startup-delay-us` definitely should stay even with the regulator-always-on. No need to remove it. I suspect removing it wasn't intended though. :)

Comment (person4268) [reply arch/arm64/boot/dts/qcom/apq8016-samsung-gtelwifiue.dts:78]: Yep, had some fun fixing a merge conflict (and it considered these two similar nodes to be part of the same thing) and I guess I missed this.

File: arch/arm64/boot/dts/qcom/apq8016-samsung-gtelwifiue.dts (hunk 10 above)
Comment (person4268) [reply arch/arm64/boot/dts/qcom/apq8016-samsung-gtelwifiue.dts]: It apparently does, so i changed it back

File: arch/arm64/boot/dts/qcom/apq8016-samsung-gtelwifiue.dts (hunk 9 above)
Comment (person4268) [reply arch/arm64/boot/dts/qcom/apq8016-samsung-gtelwifiue.dts:77]: .. And i made it even worse, time to fix

**Comment by person4268** [comment]
> I'd also appreciate if the commit adding the touch was the first to be applied (as it's more "ready" compared to the other ones) and reworded in a way as if the correct compatible is used already (I think I will add it to the driver as I already have some pending series on the mailing list, so then we'll just drop the extra compatible)

Done.

File: arch/arm64/boot/dts/qcom/apq8016-samsung-gtelwifiue.dts [hunk 11]
```diff
@@ -244,4 +280,28 @@
...
+
+	tsp_int_default: tsp-int-default {
+		pins = "gpio13";
+		function = "gpio";
+
+		drive-strength = <2>;
+		bias-disable;
+	};
+
+	reg_lcd_en_default: reg-lcd-en-default {
```
Comment (Minecrell) [inline arch/arm64/boot/dts/qcom/apq8016-samsung-gtelwifiue.dts:342]: This should be in the display commit :sweat_smile:

Comment (person4268) [reply arch/arm64/boot/dts/qcom/apq8016-samsung-gtelwifiue.dts]: WHYYY
i think i fixed it

File: arch/arm64/boot/dts/qcom/apq8016-samsung-gtelwifiue.dts [hunk 12]
```diff
@@ -46,6 +46,39 @@
...
+
+		gpio = <&msmgpio 73 GPIO_ACTIVE_HIGH>;
+		enable-active-high;
+
+		pinctrl-names = "default";
+		pinctrl-0 = <&reg_tsp_en_default>;
+
+		startup-delay-us = <150>;
+	};
+
```
Comment (TravMurav) [inline arch/arm64/boot/dts/qcom/apq8016-samsung-gtelwifiue.dts:64]: > ERROR: trailing whitespace

Comment (person4268) [reply arch/arm64/boot/dts/qcom/apq8016-samsung-gtelwifiue.dts]: hopefully that's the last thing and i didn't somehow mess up the commit history again, fixed

**Review by TravMurav — APPROVED** [review]

**Review by Minecrell — APPROVED** [review]
